### PR TITLE
Api add support bulk pause resume

### PR DESCRIPTION
**Pull request by megan-parker**
Endpoint for Bulk Updates to DAGs
---
Currently the Airflow REST API only allows us to pause/resume (update) one DAG per request. We often need to pause/resume DAGs in bulk. To avoid maintaining a separate set of custom endpoints that will be fragile, it makes sense to update the `patch_dag` endpoint to handle bulk requests.

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 1]
```diff
@@ -86,10 +86,13 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
 
 @security.requires_access([(permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG)])
 @provide_session
-def patch_dag(session, dag_id, update_mask=None):
+def patch_dag(session, dag_id, update_mask=None, multi=False):
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: This `multi` flag does not looks right to me. Bulk operations need their own endpoint instead IMO.

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_endpoint.py]: Maybe a `POST` to `/dags/~/patch` like `get_dag_runs_batch`, or just send a `PATCH` to `/dags/` would be a good idea. Not sure. But this is not the right solution.

Comment (SamWheating) [reply airflow/api_connexion/endpoints/dag_endpoint.py]: I think that this sounds like a good solution, but I would suggest that we still include a `dag_id_pattern` argument for filtering DAG IDs (which would then be consistent with the `GET /dags` endpoint)

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_endpoint.py]: Agreed, an argument to filter the list endpoint sounds like a good thing to have on its own.

Comment (megan-parker) [reply airflow/api_connexion/endpoints/dag_endpoint.py]: Thanks for the suggestion, I agree that `patch_dags` should follow the same pattern as `get_dags`.

File: airflow/api_connexion/openapi/v1.yaml [hunk 2]
```diff
@@ -441,6 +441,46 @@ paths:
...
 
+    patch:
+      summary: Update DAGs
+      x-openapi-router-controller: airflow.api_connexion.endpoints.dag_endpoint
+      operationId: patch_dags
+      tags: [DAG]
+      parameters:
+        - $ref: '#/components/parameters/PageLimit'
+        - $ref: '#/components/parameters/PageOffset'
+        - $ref: '#/components/parameters/OrderBy'
```
Comment (SamWheating) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
```

Since the order_by arg isn't included in the function.

File: airflow/api_connexion/openapi/v1.yaml [hunk 3]
```diff
@@ -441,6 +441,46 @@ paths:
...
+        - name: only_active
+          in: query
+          schema:
+            type: boolean
+            default: true
+          required: false
+          description: |
+            Only return active DAGs.
+
+            *New in version 2.1.1*
```
Comment (SamWheating) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
            Only update active DAGs.
```

File: airflow/api_connexion/openapi/v1.yaml [hunk 4]
```diff
@@ -441,6 +441,46 @@ paths:
...
+        - name: dag_id_pattern
+          in: query
+          schema:
+            type: string
+          required: false
+          description: |
+            If set, only return DAGs with dag_ids matching this pattern.
+
+            *New in version 2.3.0*
```
Comment (SamWheating) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
            If set, only update DAGs with dag_ids matching this pattern.
```

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 5]
```diff
@@ -108,6 +108,48 @@ def patch_dag(session, dag_id, update_mask=None):
...
+    """Patch all DAGs."""
+    if only_active:
+        dags_query = session.query(DagModel).filter(~DagModel.is_subdag, DagModel.is_active)
+    else:
+        dags_query = session.query(DagModel).filter(~DagModel.is_subdag)
+
+    if dag_id_pattern:
+        dags_query = dags_query.filter(DagModel.dag_id.ilike(f'%{dag_id_pattern}%'))
+
+    readable_dags = current_app.appbuilder.sm.get_accessible_dag_ids(g.user)
```
Comment (SamWheating) [inline airflow/api_connexion/endpoints/dag_endpoint.py]:
```suggestion
    editable_dags = current_app.appbuilder.sm.get_editable_dag_ids(g.user)
```

Since `get_accessible_dag_ids` returns any DAGs which are editable _or_ readable.

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 6]
```diff
@@ -108,6 +108,48 @@ def patch_dag(session, dag_id, update_mask=None):
     session.commit()
     return dag_schema.dump(dag)
 
+@security.requires_access([(permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG)])
+@format_parameters({'limit': check_limit})
+@provide_session
+def patch_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_pattern=None, update_mask=None):
+    """Patch all DAGs."""
```
Comment (SamWheating) [inline airflow/api_connexion/endpoints/dag_endpoint.py]:
```suggestion
    """Patch multiple DAGs."""
```

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 7]
```diff
@@ -108,6 +108,48 @@ def patch_dag(session, dag_id, update_mask=None):
...
+    if update_mask:
+        patch_body_ = {}
+        if len(update_mask) > 1:
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        update_mask = update_mask[0]
+        if update_mask != 'is_paused':
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        patch_body_[update_mask] = patch_body[update_mask]
+        patch_body = patch_body_
+    for dag in dags:
```
Comment (SamWheating) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: Would it be possible to do this in a single operation rather than a `for` loop? I know that mySQL doesn't support `update .. returning` queries, but maybe we could update all of the DAGs in a `query.update()` and then fetch the list separately?

I wanna say we can just do something like:
```python
dags_query = dags_query.order_by(DagModel.dag_id).offset(offset).limit(limit)

dags_query.update({DagModel.is_paused: is_paused}, synchronize_session='fetch')

dags = dags_query.all()
```

But I don't think that the set of DAGs updated will necessarily be the same as the set of DAGs returned? I guess it depends on the isolation level of the database 🤔 

Anyways, let me know if you have any suggestions, otherwise I think it's fine to leave this as an O(n) operation.

Comment (SamWheating) [reply airflow/api_connexion/endpoints/dag_endpoint.py]: Oh - maybe we can just fetch the list of DAGs to-be-paused and then mass-update them by filtering on the set?

```python
dags = dags_query.order_by(DagModel.dag_id).offset(offset).limit(limit).all()

dags_to_pause = {dag.dag_id for dag in dags}

# and then later:
session.query(DagModel).filter(DagModel.dag_id.in_(dags_to_pause)).update({DagModel.is_paused: is_paused}, synchronize_session='fetch')
```

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 8]
```diff
@@ -109,6 +109,53 @@ def patch_dag(session, dag_id, update_mask=None):
...
+        dags_query = dags_query.filter(DagModel.dag_id.ilike(f'%{dag_id_pattern}%'))
+
+    editable_dags = current_app.appbuilder.sm.get_editable_dag_ids(g.user)
+
+    dags_query = dags_query.filter(DagModel.dag_id.in_(editable_dags))
+    if tags:
+        cond = [DagModel.tags.any(DagTag.name == tag) for tag in tags]
+        dags_query = dags_query.filter(or_(*cond))
+
+    total_entries = len(dags_query.all())
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: Uh. I noticed `get_dags` also does this. It’s slow and we should use `func.count()` instead. Could you change this?

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 9]
```diff
@@ -109,6 +109,53 @@ def patch_dag(session, dag_id, update_mask=None):
...
+        raise BadRequest("Invalid Dag schema", detail=str(err.messages))
+    if update_mask:
+        patch_body_ = {}
+        if len(update_mask) > 1:
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        update_mask = update_mask[0]
+        if update_mask != 'is_paused':
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        patch_body_[update_mask] = patch_body[update_mask]
+        patch_body = patch_body_
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: This block should be moved to the beginning of the endpoint, so an invalid payload would not need to trigger database queries unnecessarily.

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 10]
```diff
@@ -109,6 +109,53 @@ def patch_dag(session, dag_id, update_mask=None):
...
+    dags_to_update = {dag.dag_id for dag in dags}
+    session.query(DagModel).filter(DagModel.dag_id.in_(dags_to_update)).update(
+        {DagModel.is_paused: patch_body['is_paused']}, synchronize_session='fetch'
+    )
+
+    session.commit()
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: This can be a `session.flush()` and delay the commit (which is done after the function finishes, by `provide_session`).

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 11]
```diff
@@ -109,6 +109,53 @@ def patch_dag(session, dag_id, update_mask=None):
...
+        patch_body = dag_schema.load(request.json, session=session)
+    except ValidationError as err:
+        raise BadRequest("Invalid Dag schema", detail=str(err.messages))
+    if update_mask:
+        patch_body_ = {}
+        if len(update_mask) > 1:
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        update_mask = update_mask[0]
+        if update_mask != 'is_paused':
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: Can this repeated check be simplied to this?

```python
if update_mask != ["is_paused"]:
    raise ...
```

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 12]
```diff
@@ -88,25 +88,68 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
...
+        patch_body_ = {}
+        if update_mask != ['is_paused']:
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        update_mask = update_mask[0]
+        patch_body_[update_mask] = patch_body[update_mask]
+        patch_body = patch_body_
     dag = session.query(DagModel).filter(DagModel.dag_id == dag_id).one_or_none()
     if not dag:
         raise NotFound(f"Dag with id: '{dag_id}' not found")
+    setattr(dag, 'is_paused', patch_body['is_paused'])
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]: Any reason why this is not

```suggestion
    dag.is_paused = patch_body['is_paused']
```

instead?

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 13]
```diff
@@ -77,7 +77,7 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
         cond = [DagModel.tags.any(DagTag.name == tag) for tag in tags]
         dags_query = dags_query.filter(or_(*cond))
 
-    total_entries = len(dags_query.all())
+    total_entries = dags_query.count()
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_endpoint.py:92]: This may give us issues later(not sure) see https://docs.sqlalchemy.org/en/14/faq/sessions.html#faq-query-deduplicating

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 14]
```diff
@@ -88,25 +88,68 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
...
+        patch_body = patch_body_
     dag = session.query(DagModel).filter(DagModel.dag_id == dag_id).one_or_none()
     if not dag:
         raise NotFound(f"Dag with id: '{dag_id}' not found")
+    setattr(dag, 'is_paused', patch_body['is_paused'])
+    session.flush()
+    return dag_schema.dump(dag)
+
+
+@security.requires_access([(permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG)])
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_endpoint.py]:
```suggestion
@security.requires_access([(permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG)])
```

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 15]
```diff
@@ -88,25 +88,68 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
 @provide_session
 def patch_dag(session, dag_id, update_mask=None):
     """Update the specific DAG"""
+    try:
+        patch_body = dag_schema.load(request.json, session=session)
+    except ValidationError as err:
+        raise BadRequest("Invalid Dag schema", detail=str(err.messages))
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_endpoint.py]:
```suggestion
        raise BadRequest(detail=str(err.messages))
```
For consistency

File: airflow/api_connexion/openapi/v1.yaml [hunk 16]
```diff
@@ -441,6 +441,41 @@ paths:
...
+          required: false
+          description: |
+            Only update active DAGs.
+        - name: dag_id_pattern
+          in: query
+          schema:
+            type: string
+          required: false
+          description: |
+            If set, only update DAGs with dag_ids matching this pattern.
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]: We should have these schemas in the components/parameters section so it can be reused if need arises

Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml]: RequestBody is needed too

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]

File: airflow/api_connexion/endpoints/dag_endpoint.py (hunk 13 above)
Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_endpoint.py:92]: I don’t think this is an issue here because `dags_query` does not have things like JOIN and shouldn’t need to be deduplicated. And even if it is, we arguably _should_ add deduplication logic to `dags_query` (otherwise offset and limit won’t work right), which will make `count()` correct in the end.

File: airflow/api_connexion/openapi/v1.yaml [hunk 17]
```diff
@@ -445,27 +428,22 @@ paths:
...
-            If set, only update DAGs with dag_ids matching this pattern.
+        - $ref: '#/components/parameters/OnlyActive'
+        - $ref: '#/components/parameters/DagIdPattern'
+      tags: [DAG]
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/DAGCollection'
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
              $ref: '#/components/schemas/DAG'
```

Comment (megan-parker) [reply airflow/api_connexion/openapi/v1.yaml]: @ephraimbuddy I made this change, but am wondering if you can clarify why the request body contains a DAG or DAGCollection schema? It doesn't seem like there is a DAG schema being passed in the request.

Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml]: Only `is_paused` is allowed to be updated but because of how clients work, we have to give the DAG object so that we generate a correct request body in clients. For example, we don't want to just have `is_paused` object, when `is_paused` is part of DAG and we already have a DAG object. The API doc knows to generate it in documentation correctly by filtering out readonly fields.

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 18]
```diff
@@ -88,25 +88,67 @@ def get_dags(limit, session, offset=0, only_active=True, tags=None, dag_id_patte
...
+    try:
+        patch_body = dag_schema.load(request.json, session=session)
+    except ValidationError as err:
+        raise BadRequest(detail=str(err.messages))
+    if update_mask:
+        patch_body_ = {}
+        if update_mask != ['is_paused']:
+            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
+        update_mask = update_mask[0]
+        patch_body_[update_mask] = patch_body[update_mask]
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_endpoint.py]:
```suggestion
        if update_mask != ['is_paused']:
            raise BadRequest(detail="Only `is_paused` field can be updated through the REST API")
        patch_body_[update_mask[0]] = patch_body[update_mask[0]]
```

Pre-emptively fix the Mypy issue that eventually needs to be fixed by the ongoing Mypy-enabling effort 🙂

File: airflow/api_connexion/openapi/v1.yaml [hunk 19]
```diff
@@ -412,25 +412,38 @@ paths:
...
+      summary: Update DAGs
+      x-openapi-router-controller: airflow.api_connexion.endpoints.dag_endpoint
+      operationId: patch_dags
+      parameters:
+        - $ref: '#/components/parameters/PageLimit'
+        - $ref: '#/components/parameters/PageOffset'
+        - $ref: '#/components/parameters/FilterTags'
+        - $ref: '#/components/parameters/UpdateMask'
+        - $ref: '#/components/parameters/OnlyActive'
+        - $ref: '#/components/parameters/DagIdPattern'
```
Comment (ashb) [inline airflow/api_connexion/openapi/v1.yaml]: This endpoint by default operates on _all_ DAGS by default right? That feels like a bit of a foot gun, and I _think_ I'd like to make dag id pattern required here, i.e. `?dag_id_pattern=%` is required to operate on all dags.

WDYT?

Comment (megan-parker) [reply airflow/api_connexion/openapi/v1.yaml]: Yes, I agree, I can add that fix in this PR. WDYT about setting `dag_id_pattern` as `~` to return all dags if `dag_id_pattern` is required?

Comment (ashb) [reply airflow/api_connexion/openapi/v1.yaml]: So long as it's documented, sounds good

Comment (ashb) [reply airflow/api_connexion/openapi/v1.yaml]: This only applies to the bulk patch endpoint, others can stay as they are

Comment (megan-parker) [reply airflow/api_connexion/openapi/v1.yaml]: Does that mean that the bulk `get` endpoint should continue to treat `dag_pattern_id` as optional and default to returning all dags if no pattern is provided?

Comment (ashb) [reply airflow/api_connexion/openapi/v1.yaml]: Yeah, the behaviour of bulk `get` shouldn't change.

File: airflow/api_connexion/openapi/v1.yaml (hunk 16 above)
Comment (megan-parker) [reply airflow/api_connexion/openapi/v1.yaml]: @ashb I had modified both the `get` and `patch` bulk endpoints. `~` can be supplied to return all dags if needed. 
I had done it this way since `dag_id_parameter` is a reusable parameter as per this thread. I am thinking to define `dag_id_pattern` separately in both bulk endpoints and define it as required in `patch` and optional in `get`. WDYT?

**Comment by megan-parker** [comment]
@ephraimbuddy could you please review again?

File: airflow/api_connexion/endpoints/dag_endpoint.py [hunk 20]
```diff
@@ -100,25 +100,67 @@ def get_dags(
...
-    setattr(dag, 'is_paused', patch_body['is_paused'])
-    session.commit()
-    return dag_schema.dump(dag)
+    if only_active:
+        dags_query = session.query(DagModel).filter(~DagModel.is_subdag, DagModel.is_active)
+    else:
+        dags_query = session.query(DagModel).filter(~DagModel.is_subdag)
+
+    if dag_id_pattern == '~':
+        dag_id_pattern = '%'
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_endpoint.py:143]: What happens when the dag_id_pattern is None? Will it update all DAGs or it won't? 
I'll still do a real world testing though, just curious

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_endpoint.py:143]: The current behaviour is to update all DAGs. So yeah this new logic seems wrong in that case, good catch.

Comment (megan-parker) [reply airflow/api_connexion/endpoints/dag_endpoint.py:143]: `dag_id_pattern` should never be `None` since it is a required parameter. 
When I pass `None` I get the following response: `Dag with id: ''None'' not found`.

File: airflow/api_connexion/openapi/v1.yaml [hunk 21]
```diff
@@ -412,25 +412,50 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DAGCollection'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
 
-            *New in version 2.1.1*
+    patch:
+      summary: Update DAGs
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml:430]: Add a description: field and also update that this is new in 2.3.0

File: airflow/api_connexion/openapi/v1.yaml [hunk 22]
```diff
@@ -412,25 +412,50 @@ paths:
...
+        - name: dag_id_pattern
           in: query
           schema:
-            type: boolean
-            default: true
+            type: string
           required: false
           description: |
-            Only return active DAGs.
+            If set, only return DAGs with dag_ids matching this pattern.
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml:418]: We should remove dag_id_pattern on /dags endpoint and make it a separate PR then add tests for it. Currently, there's no test for it and it will be good to make this change in a separate PR

File: airflow/api_connexion/openapi/v1.yaml [hunk 23]
```diff
@@ -412,25 +412,50 @@ paths:
...
+        '200':
+          description: Success.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DAGCollection'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
 
-            *New in version 2.1.1*
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml:416]: This shouldn't be removed or am I missing something? Looks like it should be at 3814 or in a description field under `/dags`

File: airflow/api_connexion/openapi/v1.yaml [hunk 24]
```diff
@@ -412,25 +412,50 @@ paths:
...
+        - $ref: '#/components/parameters/UpdateMask'
+        - $ref: '#/components/parameters/OnlyActive'
         - name: dag_id_pattern
           in: query
           schema:
             type: string
-          required: false
+          required: true
           description: |
             If set, only return DAGs with dag_ids matching this pattern.
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
            If set, only update DAGs with dag_ids matching this pattern.
```

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
Overall, it looks good. Yea, the dag_id_pattern is required on the `patch endpoint` but not required on `get endpoint`. I suggest to only work on the `patch endpoint` in this PR and work on the `get endpoint` in a separate PR.

File: airflow/api_connexion/openapi/v1.yaml (hunk 22 above)
Comment (megan-parker) [reply airflow/api_connexion/openapi/v1.yaml:418]: I hadn't added the `dag_id_pattern`, this was added by @SamWheating in [this](https://github.com/apache/airflow/pull/18924/files) PR.
I can see that the field is existing in the [main](https://github.com/apache/airflow/blob/a4d514bd7f7777e5bf03e9fa8e567f1c8774a74d/airflow/api_connexion/openapi/v1.yaml#L417) branch. It looks like the git diff is [misleading](https://github.com/apache/airflow/blame/bc8d9ba119ea7112a25d67aa70a6ea6e5069052b/airflow/api_connexion/openapi/v1.yaml#L412).

File: airflow/api_connexion/openapi/v1.yaml [hunk 25]
```diff
@@ -404,25 +408,56 @@ paths:
...
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
 
-            *New in version 2.1.1*
+    patch:
+      summary: Update DAGs
+      description: >
+       Update DAGs of a given dag_id_pattern using UpdateMask.
+
+       This endpoint allows specifying `~` as the dag_id_pattern to retrieve DAG runs for all DAGs.
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
       This endpoint allows specifying `~` as the dag_id_pattern to update all DAGs.
```

**Review by ephraimbuddy — COMMENTED** [review]
LGTM. 
One more ask, add a test to validate that the dag_id_pattern is required for this endpoint to avoid someone removing it in the future

**Review by ephraimbuddy — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
